File: 6. CountingSort.py
def counting_sort(arr):
    if not arr:
        return

    # --- Step 1: Find the maximum element to determine the range ---
    max_val = max(arr)

    # --- Step 2: Create the count array ---
    # Stores the frequency of each element.
    count = [0] * (max_val + 1)
    for num in arr:
        count[num] += 1
        
    print(f"Count array: {' '.join(map(str, count))}")

    # --- Step 3: Create the cumulative frequency array ---
    # Stores the count of elements less than or equal to the index.
    # This determines the final position of elements in the sorted array.
    cum_freq = list(count)
    for i in range(1, len(cum_freq)):
        cum_freq[i] += cum_freq[i - 1]

    print(f"Cumulative frequency array: {' '.join(map(str, cum_freq))}")

    # --- Step 4: Build the sorted array ---
    # Iterate from the end of the original array to maintain stability.
    sorted_arr = [0] * len(arr)
    for i in range(len(arr) - 1, -1, -1):
        num = arr[i]
        # Get the correct position from the cumulative frequency array
        position = cum_freq[num] - 1
        sorted_arr[position] = num
        # Decrement the count for the next occurrence of this number
        cum_freq[num] -= 1

    # --- Step 5: Copy the sorted array back to the original ---
    # This modifies the list in-place for the caller.
    arr[:] = sorted_arr

def main():
    """Main function to test the counting_sort."""
    arr = [4, 2, 2, 8, 3, 3, 1]
    
    print(f"Original array: {' '.join(map(str, arr))}")
    print("-" * 30)
    
    counting_sort(arr)
    
    print("-" * 30)
    print(f"Sorted array: {' '.join(map(str, arr))}")

if __name__ == "__main__":
    main()
    """
    Sorts a list of non-negative integers in place using Counting Sort.

    Type: Non-comparison-based, Stable

    Time Complexity:
    - Best, Average, Worst: O(n + k) where n is the number of elements
      and k is the range of the input values (max_val).

    Space Complexity:
    - Auxiliary Space: O(k) for the count/cumulative array.
    """

File: 6. CountingSort.cpp
/* 6. Counting Sort

Implement Counting Sort for non-negative integers.

Display the count array and cumulative frequency array after each step. */

#include <bits/stdc++.h>
using namespace std;

// Function to find maximum element
int findMax(vector<int> &arr) {
    int maxVal = arr[0];
    for(int num : arr)
        if(num > maxVal) maxVal = num;
    return maxVal;
}

// Function to build count array
vector<int> buildCountArray(vector<int> &arr, int maxVal) {
    vector<int> count(maxVal + 1, 0);
    for(int num : arr)
        count[num]++;
    
    cout << "Count array: ";
    for(int c : count) cout << c << " ";
    cout << endl;
    return count;
}

// Function to build cumulative frequency array
vector<int> buildCumulativeArray(vector<int> &count) {
    vector<int> cumFreq = count;
    for(size_t i = 1; i < cumFreq.size(); i++)
        cumFreq[i] += cumFreq[i-1];
    
    cout << "Cumulative frequency array: ";
    for(int c : cumFreq) cout << c << " ";
    cout << endl;
    return cumFreq;
}

// Counting Sort function
void countingSort(vector<int> &arr) {
    int maxVal = findMax(arr);
    vector<int> count = buildCountArray(arr, maxVal);
    vector<int> cumFreq = buildCumulativeArray(count);

    vector<int> sorted(arr.size());
    for(int i = arr.size()-1; i >= 0; i--) {
        sorted[cumFreq[arr[i]] - 1] = arr[i];
        cumFreq[arr[i]]--;
    }

    arr = sorted;
}

int main() {
    vector<int> arr = {4, 2, 2, 8, 3, 3, 1};
    cout << "Original array: ";
    for(int num : arr) cout << num << " ";
    cout << endl;

    countingSort(arr);

    cout << "Sorted array: ";
    for(int num : arr) cout << num << " ";
    cout << endl;

    return 0;
}

/*
Type: Non-comparison-based, Stable

Time Complexity:
Best Case: O(n + k) → Always depends on the number of elements (n) and the range of input (k)
Average Case: O(n + k)
Worst Case: O(n + k)

Space Complexity:
Auxiliary Space: O(k) → Uses count array of size equal to the range of input elements

*/